pkg/driver/node_server.go

```diff
@@ -1630,6 +1640,14 @@ func writeStagedDeviceInfo(targetPath string, stagingDev *StagingDevice) error {
 	if err != nil {
 		return err
 	}
+
+	// Attempt create of staging dir, as CSI attacher can remove the directory
```

Comment:
I don't see where the CSI attacher ever "removes" the staging directory.  I believe kubelet is doing that.  Doesn't the kubelet retry recreate the staging path?

Reply:
@rgcostea  its done [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L267) So on every failure(abort), its cleaning up. After device setup/mount is done, node plugin fails to write deviceInfo as directory is gone. I think kubelet is not creating the path again, but attacher does it with next retry. is isVolumeStaged() returns false, we end up doing staging again. If we create the directory ourselves, then attacher will again recreate vol_data.json in the parent directory and we just return with success. If this is not done, then during staging, i have to remove cleanup code if device is already attached(GST fixes), or come up with other ways to figure out if its really stale device. Let me know whichever is better.

Errors i was seeing with multiple stage requests, when directory gets deleted by attacher:

```
time="2019-11-24T22:18:34Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-6d594ae8-f272-47eb-aa2d-3ed1d61ebdc1/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-6d594ae8-f272-47eb-aa2d-3ed1d61ebdc1/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:18:34Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bff, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-6d594ae8-f272-47eb-aa2d-3ed1d61ebdc1/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:18:41Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-01287a32-7d0f-48e1-9da1-f62d71f79429/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-01287a32-7d0f-48e1-9da1-f62d71f79429/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:18:41Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bdb, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-01287a32-7d0f-48e1-9da1-f62d71f79429/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:18:48Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-ac72191b-d89c-4a8e-b161-99549968594d/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-ac72191b-d89c-4a8e-b161-99549968594d/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:18:48Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bfa, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-ac72191b-d89c-4a8e-b161-99549968594d/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:18:56Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-20fe1fa2-1ac9-4806-8642-d0b81e13a19d/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-20fe1fa2-1ac9-4806-8642-d0b81e13a19d/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:18:56Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000be4, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-20fe1fa2-1ac9-4806-8642-d0b81e13a19d/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:19:03Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-4e3044a6-3203-4a4d-aae8-c3d1867c2220/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-4e3044a6-3203-4a4d-aae8-c3d1867c2220/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:19:03Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bf3, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-4e3044a6-3203-4a4d-aae8-c3d1867c2220/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:19:10Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-801c5ad2-a4d1-4d3f-8dfd-abfb2b6ff003/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-801c5ad2-a4d1-4d3f-8dfd-abfb2b6ff003/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:19:10Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bd5, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-801c5ad2-a4d1-4d3f-8dfd-abfb2b6ff003/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:19:17Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-b4322b27-0e20-4fb5-96a9-db8b21592fd8/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-b4322b27-0e20-4fb5-96a9-db8b21592fd8/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:19:17Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000c01, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-b4322b27-0e20-4fb5-96a9-db8b21592fd8/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
time="2019-11-24T22:19:25Z" level=error msg="Failed to write to file /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66b998fe-6937-4661-82a5-38cc335b232a/globalmount/deviceInfo.json, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66b998fe-6937-4661-82a5-38cc335b232a/globalmount/deviceInfo.json: no such file or directory" file="node_server.go:1647"
time="2019-11-24T22:19:25Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to stage volume 065d1e3b28035f5f6c000000000000000000000bd8, err: open /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66b998fe-6937-4661-82a5-38cc335b232a/globalmount/deviceInfo.json: no such file or directory" file="utils.go:67"
```

Reply:
Right.  That's in kubernetes/kubernetes.  When you referred to the csi-attacher above, I thought you were referring to kubernetes-csi/external-attacher.  What you refer to above is part of kubelet.  If you look at [line 251](https://github.com/kubernetes/kubernetes/blob/9905a33fd738a52d3f7ae73cf823d82a05720e33/pkg/volume/csi/csi_attacher.go#L251) above, you'll see the attempt to create the directory again.  Is that failing?

Reply:
ah ok. i confused csi_attacher with external one. Its creating again during retry, but sometimes i think this was entering loop.

create directory -> stage request -> device attach -> retry staging -> abort(already pending) -> cleanup directory --> write deviceInfo failure --> create directory again --> stage request.

This is seen when staging requests take time to complete. If we don't create the directory, during next request, since device was already setup previously we could have just returned success with writing deviceInfo.json. But now i have added code to cleanup device on every stage request(assuming its stale). This it was causing device cleanup/setup again.

Do you think we can change cleanup/setup logic to identify only if its stale(like all failed paths, lun disconnected etc?)

Reply:
> Do you think we can change cleanup/setup logic to identify only if its stale(like all failed paths, lun disconnected etc?)

We could.  But if you think the current approach handles all cases, why bother.

Reply:
Sure, yes, if we can create the directory then next retry is gracefully handled as deviceInfo is already present.